std/json.c3

```diff
@@ -0,0 +1,428 @@
+    node.children:store(cursor.allocator:make_vector<JSON_Node>());
+
+    while true {
+        cursor:eat_whitespace();
+
+        let string_node : Optional<JSON_Node> = cursor:parse_string();
+
+        if !string_node:has_value() {
+            node.children:data():deallocate();
+            return make_optional<JSON_Node>();
+        }
+
+        node.children:data():push_back(string_node:data());
+
+        cursor:eat_whitespace();
+
+        if !cursor:eat(char(":")) {
+            node.children:data():deallocate();
+            return make_optional<JSON_Node>();
+        }
+
+        let value_node : Optional<JSON_Node> = cursor:parse_element();
+
+        if !value_node:has_value() {
+            node.children:data():deallocate();
+            return make_optional<JSON_Node>();
+        }
+
+        node.children:data():push_back(value_node:data());
+
+        if cursor:eat(char("}")) {
+            // We're done.
+            return make_optional(node);
+        }
+
+        if !cursor:eat(char(",")) {
+            node.children:data():deallocate();
+            return make_optional<JSON_Node>();
+        }
+    }
+
+    // Unreachable.
+    return make_optional(node);
+}
+
+function parse_array : (cursor : Cursor&) -> Optional<JSON_Node> = {
+    if !cursor:eat(char("[")) {
+        return make_optional<JSON_Node>();
+    }
+
+    let node : JSON_Node = new_json_node(/* array */ 5);
+
+    cursor:eat_whitespace();
+
+    if cursor:eat(char("]")) {
+        // Empty array.
+        return make_optional(node);
+    }
+
+    // Initialize children vector.
+    node.children:store(cursor.allocator:make_vector<JSON_Node>());
+
+    while true {
+        let value_node : Optional<JSON_Node> = cursor:parse_element();
+
+        if !value_node:has_value() {
+            return make_optional<JSON_Node>();
+        }
+
+        node.children:data():push_back(value_node:data());
+
+        if cursor:eat(char("]")) {
+            // We're done.
+            return make_optional(node);
+        }
+
+        if !cursor:eat(char(",")) {
+            node.children:data():deallocate();
+            return make_optional<JSON_Node>();
+        }
+    }
+
+    // Unreachable.
+    return make_optional<JSON_Node>();
+}
+
+function parse_value : (cursor : Cursor&) -> Optional<JSON_Node> = {
+    let optional_next_token : Optional<u8> = cursor:peak();
+    if !optional_next_token:has_value() {
+        return make_optional<JSON_Node>();
+    }
+
+    const next_token : u8 = optional_next_token:data();
+
+    if next_token == char("\"") {
+        return cursor:parse_string();
+    }
+
+    if next_token == char("{") {
+        return cursor:parse_object();
+    }
+
+    if next_token == char("[") {
+        return cursor:parse_array();
+    }
+
+    // Numbers don't have a prefix character, so just try and parse them.
+    let node : Optional<JSON_Node> = cursor:parse_number();
+    if node:has_value() {
+        return node;
+    }
```

Comment:
You can remove this branch by just returning `cursor:parse_number()` (after doing all the other checks ofc)

Reply:
Yeah fair enough